homeassistant/components/mqtt/siren.py

```diff
@@ -0,0 +1,357 @@
+    },
+).extend(MQTT_ENTITY_COMMON_SCHEMA.schema)
+
+DISCOVERY_SCHEMA = vol.All(PLATFORM_SCHEMA.extend({}, extra=vol.REMOVE_EXTRA))
+
+MQTT_SIREN_ATTRIBUTES_BLOCKED = frozenset(
+    {
+        ATTR_AVAILABLE_TONES,
+        ATTR_DURATION,
+        ATTR_TONE,
+        ATTR_VOLUME_LEVEL,
+    }
+)
+
+SUPPORTED_BASE = SUPPORT_TURN_OFF | SUPPORT_TURN_ON
+
+SUPPORTED_ATTRIBUTES = {
+    ATTR_DURATION: SUPPORT_DURATION,
+    ATTR_TONE: SUPPORT_TONES,
+    ATTR_VOLUME_LEVEL: SUPPORT_VOLUME_SET,
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up MQTT siren through configuration.yaml."""
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    await _async_setup_entity(hass, async_add_entities, config)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT siren dynamically through MQTT discovery."""
+    setup = functools.partial(
+        _async_setup_entity, hass, async_add_entities, config_entry=config_entry
+    )
+    await async_setup_entry_helper(hass, siren.DOMAIN, setup, DISCOVERY_SCHEMA)
+
+
+async def _async_setup_entity(
+    hass, async_add_entities, config, config_entry=None, discovery_data=None
+):
+    """Set up the MQTT siren."""
+    async_add_entities([MqttSiren(hass, config, config_entry, discovery_data)])
+
+
+class MqttSiren(MqttEntity, SirenEntity):
+    """Representation of a siren that can be toggled and notified using MQTT."""
+
+    _entity_id_format = ENTITY_ID_FORMAT
+    _attributes_extra_blocked = MQTT_SIREN_ATTRIBUTES_BLOCKED
+
+    def __init__(self, hass, config, config_entry, discovery_data):
+        """Initialize the MQTT siren."""
+        self._attr_name = config[CONF_NAME]
+        self._attr_should_poll = False
+        self._supported_features = SUPPORTED_BASE
+        self._attr_is_on = False
+        self._state_on = None
+        self._state_off = None
+        self._optimistic = None
+
+        self._attr_extra_state_attributes: dict[str, Any] = {}
+
+        self.target = None
+
+        MqttEntity.__init__(self, hass, config, config_entry, discovery_data)
+
+    @staticmethod
+    def config_schema():
+        """Return the config schema."""
+        return DISCOVERY_SCHEMA
+
+    def _setup_from_config(self, config):
+        """(Re)Setup the entity."""
+
+        state_on = config.get(CONF_STATE_ON)
+        self._state_on = state_on if state_on else config[CONF_PAYLOAD_ON]
+
+        state_off = config.get(CONF_STATE_OFF)
+        self._state_off = state_off if state_off else config[CONF_PAYLOAD_OFF]
+
+        if config[CONF_SUPPORT_DURATION]:
+            self._supported_features |= SUPPORT_DURATION
+            self._attr_extra_state_attributes[ATTR_DURATION] = None
```

Comment:
Why is this needed, is it not done by the base entity?

Reply:
The support flags need to be set, else the base entity will filter the service call parameters.
For sirens that do not support duration or volume level setting, these flags can be turned off (default = ON). This will remove the corresponding state attributes from the entity. The default is now SUPPORT_ON | SUPPORT_OFF. The support for tones will be set to on if a list of available tones is configured.

Reply:
The base entity does not support storing state attributes for tone, volume_level and duration at the moment.